external-crates/move/crates/move-package/src/resolution/dependency_graph.rs

```diff
@@ -1235,15 +1231,43 @@ impl DependencyGraph {
             self.write_dependencies_to_lock(*id, &mut writer)?;
         }
 
-        writer.flush()?;
-        std::mem::drop(writer);
+        let mut dependencies = None;
+        let mut dev_dependencies = None;
+        let mut packages = None;
+        if !writer.is_empty() {
+            let toml = writer.parse::<toml_edit::Document>()?;
+            if let Some(value) = toml.get("dependencies").and_then(|v| v.as_value()) {
+                dependencies = Some(value.clone());
+            }
+            if let Some(value) = toml.get("dev-dependencies").and_then(|v| v.as_value()) {
+                dev_dependencies = Some(value.clone());
+            }
+            packages = toml["move"]["package"].as_array_of_tables().cloned();
+        }
+
+        use std::io::Seek;
+        let mut lock = LockFile::new(
+            install_dir,
+            self.manifest_digest.clone(),
+            self.deps_digest.clone(),
+        )?;
+        lock.flush()?;
+        lock.rewind()?;
```

Comment:
We still unconditionally create a `Move.lock` from scratch as before. But this time, we will update that `Move.lock` with the `update_dependency_graph` function and edit in-place (right after this comment).

We rewind the lock file cursor so that it's ready to be read by the `update_dependency_graph` function. This is how all the `update_*` functions are set up to work in `schema.rs` (i.e., they expect a file handle they can read a lock string from).

Reply:
Does it make sense to just work with strings (or some other in-memory representation) everywhere instead of the file that you need to keep rewinding, then at the end, you can put it back into the file?

Reply:
I think I'd like something like that, but it's tricky. In effect we'd need to thread that string along wherever we'd like to avoid a write to the file, which is intimidating--mostly because the control flow is tricky, and the points at which various things care to commit or write to the file is spread out over that control flow (meaning the callers will then read/write anyway).

I'm happy to try rework this later, but consistent updates to the lock file path this way abstracts away from the callers and avoids passing a string around. Because one thing I think doesn't work is to have, e.g., `dependency_graph` or any other component work with strings be tempted to decide "ok since I have a string I care about, I'm writing to the lock file right here right now!". If we can have updates go through a predictable interface (lockfile schema `update*` functions) then we can find/reason about stateful effects to the lock (with `find-references`).

Again I may revisit and have these return strings, it might make sense, though there's also a chance that change will clutter up callers with needing to write to disk at various places.